api/src/lib.rs

```diff
@@ -3318,4 +3349,218 @@ ffi_export!(
+    _tree: JObject,
+    _position: jlong,
+) -> jboolean
+{
+    // Read tree
+    let tree = {
+        let t =_env.get_field(_tree, "merkleTreePointer", "J")
+            .expect("Should be able to get field merkleTreePointer");
+
+        read_raw_pointer(&_env, t.j().unwrap() as *const GingerLazyMHT)
+    };
+
+    // Call corresponding function and return result if Ok(), otherwise throw Exception
+    match tree.is_leaf_empty(_position as u32) {
+        Ok(result) => if result { JNI_TRUE } else { JNI_FALSE },
+        Err(e) => throw!(&_env, "java/lang/Exception", format!("Cannot check if position is empty: {}", e.to_string()).as_str(), JNI_FALSE)
+    }
+});
+
+ffi_export!(
+    fn Java_com_horizen_merkletreenative_InMemoryLazySparseMerkleTree_nativeAddLeaves(
+    _env: JNIEnv,
+    _tree: JObject,
+    _leaves: jobjectArray,
+) -> jobject
+{
+    //Read _leaves as array of OperationLeaf
+    let leaves_len = _env.get_array_length(_leaves)
+        .expect("Should be able to read leaves array size");
+    let mut leaves = vec![];
+
+    for i in 0..leaves_len {
+        // Read PositionLeaf object
+        let position_leaf_obj = _env.get_object_array_element(_leaves, i)
+            .expect(format!("Should be able to read elem {} of the leaves array", i).as_str());
+
+        // Read FieldElement from PositionLeaf.data
+        let field = {
+            let field_object = _env.get_field(
+                position_leaf_obj,
+                "data",
+                "Lcom/horizen/librustsidechains/FieldElement;"
+            ).expect("Should be able to get field FieldElement").l().unwrap();
+
+            let f =_env.get_field(field_object, "fieldElementPointer", "J")
+                .expect("Should be able to get field fieldElementPointer");
+
+            read_raw_pointer(&_env, f.j().unwrap() as *const FieldElement)
+        };
+
+        // Read position from PositionLeaf.position
+        let position = _env.get_field(
+            position_leaf_obj,
+            "position",
+            "J"
+        ).expect("Should be able to get field position").j().unwrap() as u32;
+
+        leaves.push(OperationLeaf::new(position, ActionLeaf::Insert, Some(*field)));
+    }
+
+    // Read tree
+    let tree = {
+        let t =_env.get_field(_tree, "merkleTreePointer", "J")
+            .expect("Should be able to get field merkleTreePointer");
+
+        read_mut_raw_pointer(&_env, t.j().unwrap() as *mut GingerLazyMHT)
+    };
+
+    // Update the tree with leaves and return the root
+    match tree.process_leaves(leaves.as_slice()) {
+        Ok(root) => return_field_element(&_env, root),
+        Err(e) => throw!(&_env, "java/lang/Exception", format!("Cannot insert leaves: {}", e.to_string()).as_str(), JObject::null().into_inner())
+    }    
+});
+
+ffi_export!(
+    fn Java_com_horizen_merkletreenative_InMemoryLazySparseMerkleTree_nativeRemoveLeaves(
+    _env: JNIEnv,
+    _tree: JObject,
+    _positions: jlongArray,
+) -> jobject
```

Comment:
Java part expect `void` as a result.
I think, we don't need to calculate the root as a last step. Only remove leafs if possible.